dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/trackedentity/HibernateTrackedEntityStore.java

```diff
@@ -588,59 +593,97 @@ private String getFromSubQueryJoinProgramOwnerConditions(TrackedEntityQueryParam
-      orgUnits.append(getChildrenQuery(params));
-    } else if (params.isOrganisationUnitMode(OrganisationUnitSelectionMode.SELECTED)) {
-      orgUnits.append(getSelectedQuery(params));
+    if (params.isOrganisationUnitMode(ALL) || getCurrentUserDetails().isSuper()) {
+      return orgUnits.toString();
     }
 
+    sqlParams.addValue("effectiveSearchScopePaths", getOrgUnitsPathArray(effectiveSearchOrgUnits));
+    sqlParams.addValue("captureScopePaths", getOrgUnitsPathArray(captureScopeOrgUnits));
+
+    orgUnits.append(
+        "and ((P.accesslevel in ('OPEN', 'AUDITED') and (OU.path like any (:effectiveSearchScopePaths))) ");
+    orgUnits.append(
+        "or (P.accesslevel in ('PROTECTED', 'CLOSED') and (OU.path like any (:captureScopePaths)))) ");
+
     return orgUnits.toString();
   }
 
   /**
    * Prepares the organisation units of the given parameters to simplify querying. Mode ACCESSIBLE
    * is converted to DESCENDANTS for organisation units linked to the search scope of the given
    * user. Mode CAPTURE is converted to DESCENDANTS too, but using organisation units linked to the
-   * user's capture scope, and mode CHILDREN is converted to SELECTED for organisation units
-   * including all their children. Mode can be DESCENDANTS, SELECTED, ALL only after invoking this
-   * method.
+   * user's capture scope. After invoking this method, org unit mode can only be either of
+   * DESCENDANTS, CHILDREN, SELECTED or ALL.
    */
-  private void handleOrganisationUnits(TrackedEntityQueryParams params) {
-    UserDetails user = getCurrentUserDetails();
-    if (params.isOrganisationUnitMode(OrganisationUnitSelectionMode.ACCESSIBLE)) {
-      params.setOrgUnits(
-          new HashSet<>(organisationUnitStore.getByUid(user.getUserEffectiveSearchOrgUnitIds())));
-      params.setOrgUnitMode(OrganisationUnitSelectionMode.DESCENDANTS);
-    } else if (params.isOrganisationUnitMode(OrganisationUnitSelectionMode.CAPTURE)) {
-      params.setOrgUnits(new HashSet<>(organisationUnitStore.getByUid(user.getUserOrgUnitIds())));
-      params.setOrgUnitMode(OrganisationUnitSelectionMode.DESCENDANTS);
+  private void handleOrganisationUnits(
+      TrackedEntityQueryParams params,
+      Set<OrganisationUnit> effectiveSearchOrgUnits,
+      Set<OrganisationUnit> captureScopeOrgUnits) {
+    if (params.isOrganisationUnitMode(ACCESSIBLE)) {
+      params.setOrgUnits(effectiveSearchOrgUnits);
+      params.setOrgUnitMode(DESCENDANTS);
+    } else if (params.isOrganisationUnitMode(CAPTURE)) {
+      params.setOrgUnits(captureScopeOrgUnits);
+      params.setOrgUnitMode(DESCENDANTS);
     }
   }
 
+  private Set<OrganisationUnit> getOrgUnitsFromUids(Set<String> uids) {
+    return new HashSet<>(organisationUnitStore.getByUid(uids));
+  }
+
+  private String[] getOrgUnitsPathArray(Set<OrganisationUnit> orgUnits) {
+    return orgUnits.stream().map(ou -> ou.getStoredPath() + "%").toArray(String[]::new);
```

Comment:
Can like wildcard characters `%` and `_` and its default escape character `\` be part of the orgUnit path? If so we need to escape them https://github.com/dhis2/dhis2-core/blob/master/dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/system/util/SqlUtils.java#L124 (I need to escape the `\` in this function).

https://www.postgresql.org/docs/current/functions-matching.html#FUNCTIONS-LIKE